sympy/solvers/ode.py

```diff
@@ -1482,6 +1491,241 @@ def _test_term(coeff, order):
+        sol = sol.subs(x, match['mobius'])
+        sol = sol.subs(x, x**match['k'])
+        e = e.subs(x, x**match['k'])
+
+        if B:
+            e1 = Integral(A/2, x)
+            e1 = exp(logcombine(e1, force=True))
+
+        sol = cancel((e/e1)*x**((-match['k']+1)/2))*sol
+        sol = Eq(func, sol)
+
+    # if sol is None then we can try for series solution
+    if sol is None:
+        raise NotImplementedError("The given ODE " + str(eq) + " cannot be solved by"
+            + " the hypergeometric method")
+
+    return sol
+
+def match_2nd_2F1_hypergeometric(I, k, sing_point, func):
+
+    from sympy import factor
+    x = func.args[0]
+    a = Wild("a")
+    b = Wild("b")
+    c = Wild("c")
+    t = Wild("t")
+    s = Wild("s")
+    alpha = Wild("alpha")
+    beta = Wild("beta")
+    gamma = Wild("gamma")
+    delta = Wild("delta")
+
+    rn = {'type':None}
+    # I0 of the standerd 2F1 equation.
+    I0 = ((a-b+1)*(a-b-1)*x**2 + 2*((1-a-b)*c + 2*a*b)*x + c*(c-2))/(4*x**2*(x-1)**2)
+    if sing_point != [0, 1]:
+        # If singular point is [0, 1] then we have standerd equation.
+        eqs = []
+        sing_eqs = [-beta/alpha, -delta/gamma, (delta-beta)/(alpha-gamma)]
+        # making equations for the finding the mobius transformation
+        for i in range(3):
+            if i<len(sing_point):
+                eqs.append(Eq(sing_eqs[i], sing_point[i]))
+            else:
+                eqs.append(Eq(1/sing_eqs[i], 0))
+        # solving above equations for the mobius transformation
+        _beta = -alpha*sing_point[0]
+        _delta = -gamma*sing_point[1]
+        _gamma = alpha
+        if len(sing_point) == 3:
+            _gamma = (_beta + sing_point[2]*alpha)/(sing_point[2] - sing_point[1])
+        mob = (alpha*x + beta)/(gamma*x + delta)
+        mob = mob.subs(beta, _beta)
+        mob = mob.subs(delta, _delta)
+        mob = mob.subs(gamma, _gamma)
+        mob = cancel(mob)
+        t = (beta - delta*x)/(gamma*x - alpha)
+        t = cancel(((t.subs(beta, _beta)).subs(delta, _delta)).subs(gamma, _gamma))
+    else:
+        mob = x
+        t = x
+
+    # applying mobius transformation in I to make it into I0.
+    I = I.subs(x, t)
+    I = I*(t.diff(x))**2
+    I = factor(I)
+    dict_I = {x**2:0, x:0, 1:0}
+    I0_num, I0_dem = I0.as_numer_denom()
+    # collecting coeff of (x**2, x), of the standerd equation.
+    dict_I0 = {x**2:(a-b)**2 - 1, x:2*((1-a-b)*c + 2*a*b), 1:c*(c-2)}
+    # collecting coeff of (x**2, x) from I0 of the given equation.
+    dict_I.update(collect(expand(cancel(I*I0_dem)), [x**2, x], evaluate=False))
+    eqs = []
+    # We are comparing the coeff of powers of different x, for finding the values of
+    # parameters of standerd equation.
+    for key in [x**2, x, 1]:
+        eqs.append(Eq(dict_I[key], dict_I0[key]))
+
+    _c = list(roots(eqs[2], c))[0]
+    if not _c.has(Symbol):
+        _c = min(list(roots(eqs[2], c)))
+    s = list(roots(eqs[0].subs(a-b, s), s))[-1]
```

Comment:
Roots are giving the solution in a different order in python 2 and python 3 and due to this test is failing. Is there any way to make it the same.

Reply:
The simplest way is to use `ordered`.

I wonder if there is a way to choose the "best" roots. There are 8 possible solutions to this system of 3 quadratic equations. Maybe it's worth computing all 8 and then working out which is simplest some how. Have you checked what the 8 solutions look like in a typical example?

Reply:
> I wonder if there is a way to choose the "best" roots. There are 8 possible solutions to this system of 3 quadratic equations. Maybe it's worth computing all 8 and then working out which is simplest some how. Have you checked what the 8 solutions look like in a typical example?

I am also thinking about this because sometimes a solution becomes very complex but for other combinations of a. b, c, solution is very simple. Maybe we will find all the solution and can use `count_ops` to classify the best solution.

Reply:
Firstly let's just see what they look like in a typical case. Can you show for a simple example what the 8 different solutions for a, b and c are?

Reply:
For this - `eq = x*(x-1)*f(x).diff(x, 2) + ((a+b+1)*x-c)*f(x).diff(x) + a*b*f(x)`
sols-
```
In [3]: dsolve(eq)
Out[3]: 
           1 - c  ┌─  ⎛a - c + 1, b - c + 1 │  ⎞       ┌─  ⎛a, b │  ⎞
f(x) = C₁⋅x     ⋅ ├─  ⎜                     │ x⎟ + C₂⋅ ├─  ⎜     │ x⎟
                 2╵ 1 ⎝       2 - c         │  ⎠      2╵ 1 ⎝ c   │  ⎠

In [3]: dsolve(eq)
Out[3]: 
       ⎛    1 - c  ┌─  ⎛1 - b, 1 - a │  ⎞       ┌─  ⎛-b + c, -a + c │  ⎞⎞  -a⋅log(x - 1)  -b⋅log(x - 1)  c⋅log(x - 1)
f(x) = ⎜C₁⋅x     ⋅ ├─  ⎜             │ x⎟ + C₂⋅ ├─  ⎜               │ x⎟⎟⋅ℯ             ⋅ℯ             ⋅ℯ            
       ⎝          2╵ 1 ⎝   2 - c     │  ⎠      2╵ 1 ⎝      c        │  ⎠⎠                  

In [4]: dsolve(eq)
Out[4]: 
       ⎛    1 - c  ┌─  ⎛1 - a, 1 - b │  ⎞       ┌─  ⎛-a + c, -b + c │  ⎞⎞  -a⋅log(x - 1)  -b⋅log(x - 1)  c⋅log(x - 1)
f(x) = ⎜C₁⋅x     ⋅ ├─  ⎜             │ x⎟ + C₂⋅ ├─  ⎜               │ x⎟⎟⋅ℯ             ⋅ℯ             ⋅ℯ            
       ⎝          2╵ 1 ⎝   2 - c     │  ⎠      2╵ 1 ⎝      c        │  ⎠⎠         

In [4]: dsolve(eq)
Out[4]: 
           1 - c  ┌─  ⎛b - c + 1, a - c + 1 │  ⎞       ┌─  ⎛b, a │  ⎞
f(x) = C₁⋅x     ⋅ ├─  ⎜                     │ x⎟ + C₂⋅ ├─  ⎜     │ x⎟
                 2╵ 1 ⎝       2 - c         │  ⎠      2╵ 1 ⎝ c   │  ⎠

In [3]: dsolve(eq)
Out[3]: 
         ⎛    c - 1  ┌─  ⎛a, b │  ⎞       ┌─  ⎛a - c + 1, b - c + 1 │  ⎞⎞  c⋅(-log(x) + log(x - 1))  -c⋅log(x - 1)
f(x) = x⋅⎜C₁⋅x     ⋅ ├─  ⎜     │ x⎟ + C₂⋅ ├─  ⎜                     │ x⎟⎟⋅ℯ                        ⋅ℯ             
         ⎝          2╵ 1 ⎝ c   │  ⎠      2╵ 1 ⎝       2 - c         │  ⎠⎠                       

In [3]: dsolve(eq)
Out[3]: 
         ⎛    c - 1  ┌─  ⎛-b + c, -a + c │  ⎞       ┌─  ⎛1 - b, 1 - a │  ⎞⎞  -a⋅log(x - 1)  -b⋅log(x - 1)  c⋅(-log(x) + log(x - 1))
f(x) = x⋅⎜C₁⋅x     ⋅ ├─  ⎜               │ x⎟ + C₂⋅ ├─  ⎜             │ x⎟⎟⋅ℯ             ⋅ℯ             ⋅ℯ                        
         ⎝          2╵ 1 ⎝      c        │  ⎠      2╵ 1 ⎝   2 - c     │  ⎠⎠      

In [3]: dsolve(eq)
Out[3]: 
         ⎛    c - 1  ┌─  ⎛-a + c, -b + c │  ⎞       ┌─  ⎛1 - a, 1 - b │  ⎞⎞  -a⋅log(x - 1)  -b⋅log(x - 1)  c⋅(-log(x) + log(x - 1))
f(x) = x⋅⎜C₁⋅x     ⋅ ├─  ⎜               │ x⎟ + C₂⋅ ├─  ⎜             │ x⎟⎟⋅ℯ             ⋅ℯ             ⋅ℯ                        
         ⎝          2╵ 1 ⎝      c        │  ⎠      2╵ 1 ⎝   2 - c     │  ⎠⎠                                                        

In [3]: dsolve(eq)
Out[3]: 
         ⎛    c - 1  ┌─  ⎛-a + c, -b + c │  ⎞       ┌─  ⎛1 - a, 1 - b │  ⎞⎞  -a⋅log(x - 1)  -b⋅log(x - 1)  c⋅(-log(x) + log(x - 1))
f(x) = x⋅⎜C₁⋅x     ⋅ ├─  ⎜               │ x⎟ + C₂⋅ ├─  ⎜             │ x⎟⎟⋅ℯ             ⋅ℯ             ⋅ℯ                        
         ⎝          2╵ 1 ⎝      c        │  ⎠      2╵ 1 ⎝   2 - c     │  ⎠⎠                                                        


```

Reply:
Some of those definitely look better than others. Can you tell which one is best at the stage of finding the roots though?

Reply:
all of them are almost equal.

Reply:
But what are the roots for a, b, and c in the 8 different cases? I suppose that a and b are interchangeable so there are really only 4 cases.

Reply:
parameters inside the hyper with coeff c1 are a, b, c for each of the solutions.